### PR TITLE
[ROCm 7.0] Add support for AMD CDNA4 and ROCm 7.0

### PR DESCRIPTION
**Pull request by M4jupitercannon**
### PR Category
Inference
### PR Types
New features
### Description
[ROCm 7.0] Add support for AMD CDNA4 and ROCm 7.0
Key changes:
- Update cmake/hip.cmake: Fix HIP_PATH and CMAKE_MODULE_PATH for ROCm 7.0
- Update cmake/rccl.cmake: Fix RCCL header include path
- Update cmake/thrust.cmake: Skip patches when ROCm has native shuffle
- Fix GPU architectures: Add gfx950, remove unsupported gfx926/gfx928/gfx936
- Fix hiprand/rocrand include paths for ROCm 7.0 directory structure
- Fix hipPointerAttribute_t.memoryType -> type API change
- Add HIPCC guards for thrust/rocprim headers in non-device code
- Use rocblas complex types instead of thrust::complex
- Create ROCm 7.0 patches for warpctc and warprnnt
- Disabled operators due to rocprim trait incompatibility: argsort, mode, randperm
Tested: Paddle compiled successfully with ROCm 7.0.0
### 是否引起精度变化
否

[Paddle-iluvatar PR] 92

### CODE REVIEW

**Comment by M4jupitercannon** [comment]
Added a coverage test for the ROCm branch in `_get_cuda_arch_flags` (`test_rocm_returns_empty_flags`) to address the Codecov missing lines.

Could one of the required `paddle/phi` approvers please review and approve this PR?
- @wanghuancoder
- @zrr1999
- @DanielSun11

/re-run all-failed

**Comment by M4jupitercannon** [comment]
## CI Status Update

All CI checks are now **green** except for two:

### 1. `Check approval` (policy gate)
This PR modifies files in `paddle/phi/` and requires one RD approval from: @wanghuancoder, @zrr1999, or @DanielSun11.

Could one of you please review and approve?

### 2. `Linux-IXUCA / Build and Test`
This failure is **not caused by this PR's code changes**. The `Paddle-iluvatar` build step fails because its patch set is out of sync with the current Paddle codebase. Specifically, three patches no longer apply cleanly:

- `cmake/external/warpctc.cmake` — patch context has shifted
- `paddle/phi/kernels/gpu/cross_entropy_kernel.cu` — patch context has shifted
- `paddle/phi/kernels/gpudnn/softmax_gpudnn.h` — patch context has shifted

These patches live in the `Paddle-iluvatar` repo (`build_paddle.sh`), not in this PR. The linked `[Paddle-iluvatar PR] 38` was merged but doesn't cover the context drift from recent upstream develop merges.

@YqGe585 @yongqiangma — could you either approve to skip the IXUCA check, or advise on the correct `Paddle-iluvatar` PR to link?

/re-run all-failed

File: cmake/third_party.cmake [hunk 1]
```diff
@@ -406,15 +406,11 @@ endif()
 
 list(APPEND third_party_deps extern_eigen3 extern_gflags extern_glog
      extern_xxhash)
-list(
-  APPEND
-  third_party_deps
-  extern_zlib
```
Comment (DongBaiYue) [inline cmake/third_party.cmake:412]: rocm都不需要extern_warpctc extern_warprnnt这两个库吗

File: paddle/phi/api/lib/tensor_utils.cc [hunk 2]
```diff
@@ -44,7 +44,8 @@ PADDLE_API phi::Place GetPlaceFromPtr(void* data) {
 #else
   hipPointerAttribute_t attr = {};
   hipError_t status = hipPointerGetAttributes(&attr, data);
-  if (status == hipSuccess && attr.memoryType == hipMemoryTypeDevice) {
+  // ROCm 7.0+ uses 'type' instead of 'memoryType'
+  if (status == hipSuccess && attr.type == hipMemoryTypeDevice) {
```
Comment (DongBaiYue) [inline paddle/phi/api/lib/tensor_utils.cc:48]: 不能破坏之前对海光的支持，需要支持ROCm 7.0-与ROCm 7.0+

File: tools/test_runner.py [hunk 3]
```diff
@@ -23,6 +23,7 @@
 from paddle.base import core
 
 sys.path.append(os.path.abspath(os.path.dirname(__file__)))
+
```
Comment (DongBaiYue) [inline tools/test_runner.py:26]: 无意义的修改

File: paddle/phi/kernels/CMakeLists.txt [hunk 4]
```diff
@@ -123,6 +123,22 @@ if(WITH_GPU
...
+  list(
+    REMOVE_ITEM
+    kernel_gpu
+    "gpu/argsort_kernel.cu"
+    "gpu/argsort_grad_kernel.cu"
+    "gpu/mode_kernel.cu"
+    "gpu/mode_grad_kernel.cu"
+    "gpu/randperm_kernel.cu")
+endif()
+
```
Comment (DongBaiYue) [inline paddle/phi/kernels/CMakeLists.txt:141]: 同，这里应该是rocm 7.0+ 特化的

**Comment by M4jupitercannon** [comment]
@DongBaiYue 已在 commit [`fbaf7b5240`](https://github.com/M4jupitercannon/Paddle/commit/fbaf7b5240) 按 review comment 完成对应修改，逐条对应如下：

- **[缺少 ROCm 版本抽象 / 建议版本宏分派]**
  - 在 `cmake/hip.cmake` 新增统一版本宏：`PADDLE_ROCM_VERSION`（来自 `HIP_VERSION`），并用于后续分派。
  - `HIP_VERSION` 无法解析时改为 `FATAL_ERROR`，避免错误回退到旧分支。
  - 同时补充 `PADDLE_AMDGPU_TARGETS`（可配置）作为统一 arch 来源。

- **[需要同时支持 ROCm 7.0- 与 7.0+]**
  - `paddle/phi/api/lib/tensor_utils.cc` 按 `PADDLE_ROCM_VERSION` 分派：
    - `>= 70000000` 使用 `attr.type`
    - `< 70000000` 使用 `attr.memoryType`
  - 并补齐 HIP pinned memory 到 `GPUPinnedPlace()` 的处理。

- **[“这里应该是 rocm 7.0+ 特化”]**
  - `paddle/phi/kernels/CMakeLists.txt` 中 `argsort/mode/randperm` 的移除逻辑改为仅在 `PADDLE_ROCM_VERSION >= 70000000` 生效，不再对所有 `WITH_ROCM` 一刀切。
  - 同步清理了失效的 `ROCM70_UNSUPPORTED_OPS.md` 注释引用。

- **[GPU 架构列表不要丢海光型号 / 需可配置]**
  - 默认策略改为“兼容优先（联合架构）”，保留并合并：
    - `gfx906;gfx926;gfx928;gfx936;gfx942;gfx950`
  - 主构建使用 `PADDLE_AMDGPU_TARGETS`；
  - `python/paddle/utils/cpp_extension/extension_utils.py` 中 `get_rocm_arch_flags()` 同步对齐，并支持 `PADDLE_ROCM_ARCH_LIST` 覆盖。

- **[ROCm 下是否不需要 warpctc/warprnnt?]**
  - `cmake/third_party.cmake` 去掉了 `if(NOT WITH_ROCM)` 对 `extern_warpctc/extern_warprnnt` 的额外排除，保持 ROCm 路径可用。
  - `cmake/external/warpctc.cmake` / `cmake/external/warprnnt.cmake` 改为按版本选择 patch：
    - `>= 7.0` 使用 `hip.cmake.rocm70`
    - `< 7.0` 使用 legacy 路径
  - 顺手修复 `warprnnt` 中 `${WARPCTC_CCBIN_OPTION}` 误用为 `${WARPRNNT_CCBIN_OPTION}`。

- **[无意义修改]**
  - `tools/test_runner.py` 的空行噪声已回退。

File: ci/rocm_test.sh [hunk 5]
```diff
@@ -0,0 +1,61 @@
+# Copyright (c) 2026 PaddlePaddle Authors. All Rights Reserved.
```
Comment (DongBaiYue) [inline ci/rocm_test.sh]: 实际没有添加CI，建议撤销CI目录下的代码修改

File: patches/hip.cmake.rocm70 [hunk 6]
```diff
@@ -0,0 +1,108 @@
+if(NOT WITH_ROCM)
```
Comment (DongBaiYue) [inline patches/hip.cmake.rocm70]: patches/warpctc/hip.cmake.rocm70与patches/warprnnt/hip.cmake.rocm70完全一致，且与主cmake风格不同。确认是否必要

File: cmake/hip.cmake [hunk 7]
```diff
@@ -6,24 +6,37 @@ if(NOT DEFINED ENV{ROCM_PATH})
...
-      CACHE PATH "Path to which clang has been installed")
 else()
   set(ROCM_PATH
       $ENV{ROCM_PATH}
       CACHE PATH "Path to which ROCm has been installed")
+endif()
+
+# ROCm 7.0+: HIP is now directly under ROCM_PATH, not in a separate hip subdirectory
+# Check if we're using newer ROCm layout (7.0+) or older layout
+if(EXISTS "${ROCM_PATH}/lib/cmake/hip/FindHIP.cmake")
```
Comment (DongBaiYue) [inline cmake/hip.cmake]: ROCm 7.0+ 和 Fallback 分支完全相同，可合并

File: patches/hip.cmake.rocm70 (hunk 6 above)
Comment (DongBaiYue) [reply patches/hip.cmake.rocm70]: 可否复用主 `cmake/hip.cmake` 的变量

File: cmake/hip.cmake [hunk 8]
```diff
@@ -93,10 +135,10 @@ endif()
 
 # set CXX flags for HIP
 set(CMAKE_C_FLAGS
-    "${CMAKE_C_FLAGS} -D__HIP_PLATFORM_HCC__ -D__HIP_PLATFORM_AMD__ -DROCM_NO_WRAPPER_HEADER_WARNING"
+    "${CMAKE_C_FLAGS} -D__HIP_PLATFORM_HCC__ -D__HIP_PLATFORM_AMD__ -D__HIP__=1 -DROCM_NO_WRAPPER_HEADER_WARNING"
```
Comment (DongBaiYue) [inline cmake/hip.cmake]: `-D__HIP__=1` 的含义？是否必要

File: cmake/thrust.cmake [hunk 9]
```diff
@@ -1,26 +1,38 @@
...
-    "#include \"thrust/shuffle.h\"\n"
-    "#include \"stdio.h\"\n"
-    "int main() {\n"
-    "  int version = THRUST_VERSION;\n"
-    "  printf(\"%d\", version);\n"
-    "  return 0;\n"
-    "}\n")
+  # ROCm 7.0+ has rocThrust with shuffle support built-in, so no patches needed
+  if(WITH_ROCM)
+    # Check if rocThrust has shuffle.h
```
Comment (DongBaiYue) [inline cmake/thrust.cmake]: 这里是ROCm 7.0+特化的？还是ROCM的？

Comment (DongBaiYue) [reply cmake/thrust.cmake]: 看了下DCU也有${ROCM_PATH}/include/thrust/shuffle.h这个文件

File: paddle/phi/common/complex.h [hunk 10]
```diff
@@ -28,8 +28,12 @@
 
 #ifdef PADDLE_WITH_HIP
 #include <hip/hip_complex.h>
+// thrust/complex.h requires hipcc compiler
+// (rocThrust 7.0+ pulls in rocprim)
```
Comment (DongBaiYue) [inline paddle/phi/common/complex.h]: 添加的宏是否必要

File: patches/warpctc/hip.cmake.rocm70 [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+list(APPEND HIP_HCC_FLAGS --amdgpu-target=gfx908)
+list(APPEND HIP_HCC_FLAGS --amdgpu-target=gfx90a)
+list(APPEND HIP_HCC_FLAGS --amdgpu-target=gfx942)
+list(APPEND HIP_HCC_FLAGS --amdgpu-target=gfx950)
+list(APPEND HIP_CLANG_FLAGS -fno-gpu-rdc)
+list(APPEND HIP_CLANG_FLAGS --amdgpu-target=gfx906)
+list(APPEND HIP_CLANG_FLAGS --amdgpu-target=gfx908)
+list(APPEND HIP_CLANG_FLAGS --amdgpu-target=gfx90a)
+list(APPEND HIP_CLANG_FLAGS --amdgpu-target=gfx942)
+list(APPEND HIP_CLANG_FLAGS --amdgpu-target=gfx950)
```
Comment (DongBaiYue) [inline patches/warpctc/hip.cmake.rocm70]: hip.cmake中`PADDLE_AMDGPU_TARGETS` 默认列表不包含 gfx908 (MI100) 和 gfx90a (MI250)，但 `hip.cmake.rocm70` 包含。

File: paddle/phi/kernels/funcs/rocprim_traits.h [hunk 12]
```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2026 PaddlePaddle Authors. All Rights Reserved.
```
Comment (DongBaiYue) [inline paddle/phi/kernels/funcs/rocprim_traits.h]: 这个修改有在算子上验证吗？argsort算子被移除cmakelist了

File: paddle/phi/common/complex.h (hunk 10 above)
Comment (DongBaiYue) [reply paddle/phi/common/complex.h]: 确认是否是ROCM 7.0+特化的？或者兼容ROCM 7.0-？

**Comment by M4jupitercannon** [comment]
@DongBaiYue 基于当前最新提交 `b85be396` 重新整理 Apr 20 / Apr 23 review comments 的对应处理，逐项回复如下：

1. `extern_warpctc` / `extern_warprnnt`：ROCm 下仍保留这两个 third-party 依赖；当前 PR 已不再修改 `cmake/third_party.cmake`，不会在 `WITH_ROCM` 下排除它们。`cmake/external/warpctc.cmake` / `cmake/external/warprnnt.cmake` 中，ROCm 7.0+ 复制并复用主工程 `cmake/hip.cmake`；ROCm 7.0- 仍走原 legacy patch 路径。同时保留了 `warprnnt` 中 `${WARPCTC_CCBIN_OPTION}` -> `${WARPRNNT_CCBIN_OPTION}` 的修正。

2. 海光 / ROCm 7.0- 兼容：已用 `PADDLE_ROCM_VERSION` 做版本分派。`paddle/phi/api/lib/tensor_utils.cc` 中 ROCm 7.0+ 使用 `attr.type`，ROCm 7.0- 使用 `attr.memoryType`，并补齐 pinned host memory 到 `GPUPinnedPlace()` 的处理。`cmake/hip.cmake` 现在优先识别 `${ROCM_PATH}/hip/cmake/FindHIP.cmake`，保留 Hygon/DCU 布局；否则使用 ROCm 7.0+ 的 `${ROCM_PATH}/lib/cmake/hip` 布局。

3. `tools/test_runner.py`：无意义空行修改已回退；当前 PR diff 中不再包含该文件。

4. `paddle/phi/kernels/CMakeLists.txt`：当前 PR 已不再修改该文件，也不再通过 CMake 移除 `argsort` / `mode` / `randperm`。后续提交已恢复 DCU kernel registration；ROCm 7.0+ 的 rocPRIM 4.x 兼容逻辑改为在相关算子实现内按 `ROCPRIM_VERSION` 局部处理，避免影响 ROCm 7.0- / DCU legacy 路径。

5. CI 目录修改：未实际接入的新 `ci/rocm_test.sh` 已删除，`ci/utils.sh` 的 ROCm runner 调整也已移除。因此当前 PR 不再包含 `ci/` 目录下的代码修改。

6. `patches/hip.cmake.rocm70`：重复 patch 已移除，当前 PR 不再新增 `patches/warpctc/hip.cmake.rocm70` 或 `patches/warprnnt/hip.cmake.rocm70`。ROCm 7.0+ 的 third-party build 直接复用主 `cmake/hip.cmake`；`ExternalProject` 透传 `ROCM_PATH`、`HIP_PATH`、`PADDLE_AMDGPU_TARGETS`。主 `cmake/hip.cmake` 也做了 third-party 场景兼容：非主工程上下文只查找 `hiprand` / `rocrand` / `rocthrust`，并对 `include(thrust)` 做存在性判断。

7. `cmake/hip.cmake` 中 ROCm 7.0+ / fallback 分支：已合并简化布局判断。当前逻辑优先 Hygon/DCU legacy layout，再 fallback 到 ROCm 7.0+ layout，不再保留重复分支。

8. `-D__HIP__=1`：已移除，只保留必要的 HIP platform 宏。

9. `cmake/thrust.cmake`：现在按 `PADDLE_ROCM_VERSION` 判断，不再通过 `shuffle.h` 是否存在来判断。ROCm 7.0+ 跳过本地 thrust patch；ROCm 7.0- / DCU 即使已有 `shuffle.h` 也仍保留旧 patch 行为；CUDA 仍保留原来的编译探测逻辑。

10. `paddle/phi/common/complex.h`：这里不是 ROCm 7.0+ 特化，而是 HIP 编译模式特化。`thrust/complex.h` 和 `thrust::complex` interop 只在 `__HIPCC__` 下启用，避免普通 host 编译器包含 rocThrust/rocprim 头导致编译问题。相关头文件也同步做了 `__HIPCC__` guard，`blas_impl.hip.h` 改为直接使用 `rocblas_*_complex`，减少对 `thrust/complex.h` 的依赖。

11. AMDGPU arch 默认值：主工程和 `cpp_extension` 已对齐。ROCm 7.0+ 默认包含 `gfx906;gfx908;gfx90a;gfx926;gfx928;gfx936;gfx942;gfx950`；ROCm 7.0- 保持 Hygon/DCU 默认集合 `gfx906;gfx926;gfx928;gfx936`。主工程可用 `PADDLE_AMDGPU_TARGETS` 覆盖，`cpp_extension` 可用 `PADDLE_ROCM_ARCH_LIST` 覆盖；third-party ROCm 7.0+ build 复用主工程传入的 arch 列表。

12. rocPRIM traits 验证：当前已不再新增独立的 `rocprim_traits.h`，而是在 `argsort_kernel.cu`、`mode.h`、`top_k_function_cuda.h` 等实际使用排序 / top-k 的位置按 `ROCPRIM_VERSION >= 400000` 局部适配 `rocprim::traits::define`，旧 ROCm 继续使用原 `rocprim::detail` traits。因为 `CMakeLists.txt` exclusion 已移除，`argsort` / `mode` 等算子路径会继续参与构建；最新一轮 CI 中 `Linux-DCU / Build` 已通过，`Linux-DCU / Test` 仍在运行中。

当前 CI 状态：`Pre Commit`、`Linux-build`、`Linux-CPU`、`Linux-XPU`、`Windows-*`、`PR-CI-Inference`、`PR-CI-SOT`、Fleet 两项等已通过；除审批门禁 `Check approval` 外，当前明确失败仍是 `Linux-IXUCA / Build and Test`，更像是 Paddle-iluvatar fork / patch 与当前 develop 不同步；`Mac-CPU` 为 cancelled，另有一个 `Coverage test` 和 `Linux-DCU / Test` 仍在 running。

**Review by wanghuancoder — APPROVED** [review]
